source/common/upstream/maglev_lb.h

```diff
@@ -62,9 +69,81 @@ class MaglevTable : public ThreadAwareLoadBalancerBase::HashingLoadBalancer,
+                                    double max_normalized_weight, bool use_hostname_for_hashing);
+
   const uint64_t table_size_;
-  std::vector<HostConstSharedPtr> table_;
   MaglevLoadBalancerStats& stats_;
+
+private:
+  /**
+   * Implementation specific construction of data structures to represent the
+   * Maglev Table.
+   * TODO(kbaichoo): make vector const?
+   * TODO(kbaichoo): make these pure
+   */
+  virtual void constructImplementationInternals(std::vector<TableBuildEntry>& table_build_entries,
+                                                double max_normalized_weight) PURE;
+
+  /**
+   * Log each entry of the maglev table (useful for debugging).
+   */
+  virtual void logMaglevTable(bool use_hostname_for_hashing) const PURE;
+};
+
+/**
+ * This is an implementation of Maglev consistent hashing that directly holds
+ * pointers.
+ */
+class OriginalMaglevTable : public MaglevTable {
+public:
+  OriginalMaglevTable(const NormalizedHostWeightVector& normalized_host_weights,
+                      double max_normalized_weight, uint64_t table_size,
+                      bool use_hostname_for_hashing, MaglevLoadBalancerStats& stats)
+      : MaglevTable(table_size, stats) {
+    constructMaglevTableInternal(normalized_host_weights, max_normalized_weight,
+                                 use_hostname_for_hashing);
+  }
+  virtual ~OriginalMaglevTable() = default;
+
+  // ThreadAwareLoadBalancerBase::HashingLoadBalancer
+  HostConstSharedPtr chooseHost(uint64_t hash, uint32_t attempt) const override;
+
+private:
+  void constructImplementationInternals(std::vector<TableBuildEntry>& table_build_entries,
+                                        double max_normalized_weight) override;
+
+  void logMaglevTable(bool use_hostname_for_hashing) const override;
+  std::vector<HostConstSharedPtr> table_;
+};
+
+/**
+ * This maglev implementation leverages the number of hosts to more efficiently
+ * populate the maglev table.
+ */
+class CompactMaglevTable : public MaglevTable {
+public:
+  CompactMaglevTable(const NormalizedHostWeightVector& normalized_host_weights,
+                     double max_normalized_weight, uint64_t table_size,
+                     bool use_hostname_for_hashing, MaglevLoadBalancerStats& stats);
+  virtual ~CompactMaglevTable() = default;
+
+  // ThreadAwareLoadBalancerBase::HashingLoadBalancer
+  HostConstSharedPtr chooseHost(uint64_t hash, uint32_t attempt) const override;
+
+private:
+  void constructImplementationInternals(std::vector<TableBuildEntry>& table_build_entries,
+                                        double max_normalized_weight) override;
+  void logMaglevTable(bool use_hostname_for_hashing) const override;
+
+  // Leverage a BitArray to more compactly fit represent the MaglevTable.
+  // The BitArray will index into the host_table_ which will provide the given
+  // host to load balance to.
+  BitArray table_;
```

Comment:
There's seems a fair amount of code duplicated between the two derived classes. 

What do you think about having an abstraction around just `table_` and `host_table_` for `CompactMaglevTable` and `table_` for `OriginalMaglevTable`?

Reply:
I was unclear with my wording. I was suggesting that instead of duplicating the algorithmic details in both derived classes, limit the functionality of the derived classes to just setting/getting the host for an index.